server/src/main/proto/spine/system/server/c/history_commands.proto

```diff
@@ -0,0 +1,192 @@
+//
+// Copyright 2018, TeamDev. All rights reserved.
+//
+// Redistribution and use in source and/or binary forms, with or without
+// modification, must retain the above copyright notice and the following
+// disclaimer.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+syntax = "proto3";
+
+package spine.system.server;
+
+import "spine/options.proto";
+
+option (type_url_prefix) = "type.spine.io";
+option (internal_all) = true;
+
+option java_package = "io.spine.system.server";
+option java_outer_classname = "HistoryCommandsProto";
+option java_multiple_files = true;
+
+import "google/protobuf/any.proto";
+
+import "spine/core/event.proto";
+import "spine/core/command.proto";
+import "spine/system/server/entity_history.proto";
+
+// A command to create a new entity.
+//
+// This command is sent when the first messag is dispatched to the entity.
+//
+message CreateEntity {
+
+    // The ID of the entity.
+    EntityHistoryId id = 1 [(required) = true];
+
+    // The kind of the entity.
+    EntityOption.Kind kind = 2 [(required) = true];
+}
+
+// A command to dispatch an event to a subscriber.
+//
+// This command is relevant only to histories of Projections.
+//
+message DispatchEventToSubscriber {
+
+    // The ID of the entity, which the event is dispatched to.
+    EntityHistoryId receiver = 1 [(required) = true];
+
+    // The ID of the event to dispatch.
+    spine.core.EventId event_id = 2 [(required) = true, (valid) = true];
+}
+
+// A command to dispatch an event to a reactor.
+//
+// This command is relevant to histories of Aggregates and ProcessManagers.
+//
+message DispatchEventToReactor {
+
+    // The ID of the entity, which the event is dispatched to.
+    EntityHistoryId receiver = 1 [(required) = true];
+
+    // The ID of the event to dispatch.
+    spine.core.EventId event_id = 2 [(required) = true, (valid) = true];
+}
+
+// A command to pass an event to an applier.
+//
+// This command is relevant only to histories of Aggregates.
+//
+// This command is not posted if applying the event to the Aggregate fails.
+//
+message PassEventToApplier {
```

Comment:
We need to state if this event is emitted every time an `Aggregate` is being read. I am not sure whether you tried to make that clear by `This command is relevant only to histories of Aggregates.`

Just in case:

To me, emitting this event upon every `findOrCreate(...)` will kill any system, as long as total number of `PassEventToApplier` in this case is probably 50x the combined amount of commands and events  dispatched to an `Aggregate`. 

Just imagine, how many system events we are going to have for an instance of BKO `Order` aggregate with 1500 images each producing ~5-6 of `Order` domain events during its lifecycle(Create, Attach, Export etc) — as most of `Order` events are the consequences of reactions to someoneelse's events, and `findOrCreate` is invoked each time.